server/src/test/java/org/elasticsearch/search/aggregations/bucket/histogram/DateRangeHistogramAggregatorTests.java

```diff
@@ -658,6 +660,118 @@ public void testWithinQuery() throws Exception {
+                new LongBounds("2019-08-02T03:00:00", "2019-08-02T10:00:00")),
+            writer -> {
+                writer.addDocument(singleton(new BinaryDocValuesField(FIELD_NAME, RangeType.DATE.encodeRanges(singleton(range1)))));
+                writer.addDocument(singleton(new BinaryDocValuesField(FIELD_NAME, RangeType.DATE.encodeRanges(singleton(range2)))));
+            },
+            histo -> {
+                assertEquals(8, histo.getBuckets().size());
+
+                assertEquals(asZDT("2019-08-02T03:00:00"), histo.getBuckets().get(0).getKey());
+                assertEquals(1, histo.getBuckets().get(0).getDocCount());
+
+                assertEquals(asZDT("2019-08-02T05:00:00"), histo.getBuckets().get(2).getKey());
+                assertEquals(2, histo.getBuckets().get(2).getDocCount());
+
+                assertEquals(asZDT("2019-08-02T10:00:00"), histo.getBuckets().get(7).getKey());
+                assertEquals(1, histo.getBuckets().get(7).getDocCount());
+
+                assertTrue(AggregationInspectionHelper.hasValue(histo));
+            }
+        );
+    }
+
+    public void testBothBounds() throws Exception {
+        RangeFieldMapper.Range range1 = new RangeFieldMapper.Range(RangeType.DATE, asLong("2019-08-02T02:15:00"),
+            asLong("2019-08-02T05:45:00"), true, true);
+        RangeFieldMapper.Range range2 = new RangeFieldMapper.Range(RangeType.DATE, asLong("2019-08-02T05:15:00"),
+            asLong("2019-08-02T17:45:00"), true, true);
+
+        testCase(
+            Queries.newMatchAllQuery(),
+            builder -> builder.calendarInterval(DateHistogramInterval.HOUR)
+                .hardBounds(new LongBounds("2019-08-02T00:00:00", "2019-08-02T10:00:00"))
+                .extendedBounds(new LongBounds("2019-08-02T01:00:00", "2019-08-02T08:00:00")),
+            writer -> {
+                writer.addDocument(singleton(new BinaryDocValuesField(FIELD_NAME, RangeType.DATE.encodeRanges(singleton(range1)))));
+                writer.addDocument(singleton(new BinaryDocValuesField(FIELD_NAME, RangeType.DATE.encodeRanges(singleton(range2)))));
+            },
+            histo -> {
+                assertEquals(10, histo.getBuckets().size());
+
+                assertEquals(asZDT("2019-08-02T01:00:00"), histo.getBuckets().get(0).getKey());
+                assertEquals(0, histo.getBuckets().get(0).getDocCount());
+
+                assertEquals(asZDT("2019-08-02T02:00:00"), histo.getBuckets().get(1).getKey());
+                assertEquals(1, histo.getBuckets().get(1).getDocCount());
+
+                assertEquals(asZDT("2019-08-02T10:00:00"), histo.getBuckets().get(9).getKey());
+                assertEquals(1, histo.getBuckets().get(9).getDocCount());
+
+                assertTrue(AggregationInspectionHelper.hasValue(histo));
+            }
+        );
+    }
+
+    public void testOverlappingBounds() throws Exception {
```

Comment:
I think it would be good to have a test for hard bounds exactly equal to extended bounds.  That's a bit of an edge case, and we want to make sure it doesn't break down the line.